### PR TITLE
feat(linter): no accumulating spread

### PR DESCRIPTION
**Pull request by DonIsaac**
Creates a new nursery-level rule, `no-reduce-spread`, that prevents spreading accumulator objects/arrays in `Array.prototype.reduce`.

The Tl;Dr of why this should be avoided is because it drastically increases time/memory complexity in reductions. In general, it turns potentially `O(1)`/`O(n)` memory operations into `O(n)`/`O(n^2)`, and `O(n)` time into `O(n^2)`. For more details, refer to [this helpful blog post](https://prateeksurana.me/blog/why-using-object-spread-with-reduce-bad-idea/).

Note that this rule doesn't exist in ESLint's default rule set or any other ESLint plugin, although it looks like [there's been some interest in it in the past](https://github.com/vercel/style-guide/issues/18). Since there is no reference implementation to compare against, and thus this could potentially have bugs, I've decided to make this a nursery-level rule until we're sure it's stable and useful.

---

Edit: 

 - [ ] Sync with Biome - https://biomejs.dev/linter/rules/no-accumulating-spread/

### CODE REVIEW

**Comment by Boshen** [comment]
I've been afk, will get to this soon.

File: crates/oxc_linter/src/rules.rs [hunk 1]
```diff
@@ -226,6 +226,10 @@ mod jsx_a11y {
     pub mod img_redundant_alt;
 }
 
+mod perf {
```
Comment (Boshen) [inline crates/oxc_linter/src/rules.rs:229]: I'll add our own `oxc` mod here.

File: crates/oxc_linter/src/rules/perf/no_reduce_spread.rs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    /// }, {})
+    /// ```
+    ///
+    /// Fail
+    /// ```javascript
+    /// arr.reduce((acc, x) => ({ ...acc, [x]: fn(x) }), {})
+    /// Object.keys(obj).reduce((acc, el) => ({ ...acc, [el]: fn(el) }), {})
+    /// ```
+    NoReduceSpread,
+    nursery
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/perf/no_reduce_spread.rs:62]: I'll change this to `perf`

**Comment by Boshen** [comment]
Testing in the vscode repo:

False positives:

```
  ⚠ perf(no-reduce-spread): Do not spread accumulators in Array.prototype.reduce()
      ╭─[src/vs/workbench/contrib/extensions/browser/extensionEditor.ts:1300:1]
 1300 │             const viewContainersForLocation: IViewContainer[] = contrib[location];
 1301 │             result.push(...viewContainersForLocation.map(viewContainer => ({ ...viewContainer, location })));
      ·                                                                              ────────────────
 1302 │             return result;
      ╰────
  help: Prefer property assignment or Array.prototype.push()
```

File: crates/oxc_linter/src/snapshots/no_reduce_spread.snap [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+   ╭─[no_reduce_spread.tsx:1:1]
+ 1 │ arr.reduce(function (acc, x) { return { ...acc, [x]: x } }, {})
+   ·                                         ──────
+   ╰────
+  help: Prefer property assignment or Array.prototype.push()
+
+  ⚠ perf(no-reduce-spread): Do not spread accumulators in Array.prototype.reduce()
+   ╭─[no_reduce_spread.tsx:1:1]
+ 1 │ arr.reduce((acc, x) => {
+ 2 │             let temp = { ...acc, x }
```
Comment (Boshen) [inline crates/oxc_linter/src/snapshots/no_reduce_spread.snap:50]: Should we allow this? I'm seeing lots of cases for this in our internal repos.

Should only target specifically the `...acc, [x]: x` case? (I only glanced over the blog posts.)

**Comment by Boshen** [comment]
@camc314 Do you have the time to clean up this rule (also sync its name and tests with Biome)? https://biomejs.dev/linter/rules/no-accumulating-spread

File: crates/oxc_linter/src/snapshots/no_reduce_spread.snap (hunk 3 above)
Comment (camc314) [reply crates/oxc_linter/src/snapshots/no_reduce_spread.snap:50]: my 2 cents: i don't think so - this can pretty easilly be rewritten as 
```
arr.reduce((acc, x) => {
            acc[x] = x
            return acc
}, 
```

which omits the spread, and avoids the `O(n^2)` time complexity - i can't think ov any advantages to spreading here 🤔

**Comment by camc314** [comment]
> @camc314 Do you have the time to clean up this rule (also sync its name and tests with Biome)? https://biomejs.dev/linter/rules/no-accumulating-spread

yes, will try to do today.

i've fixed the naming ect, will work on the falst positive - i think if we cross a function boundary, we can just break out of that loop where we check for `reduce` calls and we should be fine

File: crates/oxc_linter/src/snapshots/no_reduce_spread.snap
Comment (Boshen) [inline crates/oxc_linter/src/snapshots/no_reduce_spread.snap:1]: Let's remove this file.

**Review by Boshen — CHANGES_REQUESTED** [review]
Last changes:

Bring in the valid and invalid cases from Biome:

https://github.com/biomejs/biome/blob/main/crates/biome_js_analyze/tests/specs/performance/noAccumulatingSpread/valid.jsonc

https://github.com/biomejs/biome/blob/main/crates/biome_js_analyze/tests/specs/performance/noAccumulatingSpread/invalid.jsonc

File: crates/oxc_linter/src/rules/oxc/no_accumulating_spread.rs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    #[error("oxc(no-accumulating-spread): Do not spread accumulators in Array.prototype.reduce()")]
+    #[diagnostic(severity(warning), help("It looks like you're spreading an `Array`. Consider using the `Array.push` or `Array.concat` methods to mutate the accumulator instead.\nUsing spreads within accumulators leads to `O(n^2)` time complexity."))]
+    LikelyArray(#[label] Span),
+    #[error("oxc(no-accumulating-spread): Do not spread accumulators in Array.prototype.reduce()")]
+    #[diagnostic(severity(warning), help("It looks like you're spreading an `Object`. Consider using the `Object.assign` or assignment operators to mutate the accumulator instead.\nUsing spreads within accumulators leads to `O(n^2)` time complexity."))]
+    LikelyObject(#[label] Span),
+    #[error("oxc(no-accumulating-spread): Do not spread accumulators in Array.prototype.reduce()")]
+    #[diagnostic(severity(warning), help("Consider using `Object.assign()` or `Array.prototype.push()` to mutate the accumulator instead.\nUsing spreads within accumulators leads to `O(n^2)` time complexity."))]
+    Unknown(#[label] Span),
+}
```
Comment (camc314) [inline crates/oxc_linter/src/rules/oxc/no_accumulating_spread.rs:26]: @Boshen i updated this again, mind taking another look

**Review by Boshen — APPROVED** [review]

**Comment by Boshen** [comment]
Great work team @DonIsaac and @camc314 !
